parlai/core/torch_generator_agent.py

```diff
@@ -479,8 +480,10 @@ def __init__(self, opt: Opt, shared=None):
         else:
             # this is not a shared instance of this class, so do full init
             self.criterion = self.build_criterion()
-            # ensure all distributed copies will always be in sync
-            self.model = self.build_model()
+            with fsdp_utils.maybe_fsdp_wrap(opt):
+                self.model = fsdp_utils.fsdp_wrap(self.build_model())
+                if self.fp16 and not fsdp_utils.should_use_fsdp(opt):
```

Comment:
remember that bug with the instability stuff? is this not re-introducing it?

Reply:
(because we moved the model.half() call?)

Reply:
Okay I think this needs to use my utility `should_delay_halving`. Forgot this.

We haven't really moved it the moment of halving. The operations between these two points don't do much, and the original code path should be about the same.

- We now half it on CPU instead of GPU, and then transfer. That's probably a small speedup in initialization really, with maybe some small numerical differences
- We model parallel after halving. Probably small speedup at initialization.
- We synchronize parameters after halving. Again, small initialization speedup.

The catch is that FSDP expects the model pre-halved if we're doing safe optimization, and post-halved if we're doing memory-efficient. (Similar to the optimizer wrappers, it looks for parameters of types to decide what type are the gradients).

This is the desired pattern
- If we're in Safe and using DDP, we SHOULD still halve, just as before
- If we're in MemEff and using DDP, we SHOULD still halve, just as before
- If we're in Safe and Zero2, we should NOT halve here
- If we're in MemEff and Zero2, we SHOULD halve here.